File: Regex/Matcher.cpp
#include "Matcher.hpp"

#include "NFA.hpp"

namespace Regex {

    Matcher::Matcher(const std::vector<std::string> &patterns)
    {
        mNumPatterns = (unsigned int)patterns.size();

        std::vector<std::unique_ptr<Parser::Node>> nodes;
        
        try {
            for(unsigned int i=0; i<patterns.size(); i++) {
                const std::string &pattern = patterns[i];
                mParseError.pattern = i;
                nodes.push_back(Parser::parse(pattern));
            }
        } catch(Parser::ParseException e) {
            mParseError.character = e.pos;
            mParseError.message = e.message;
            return;
        }

        mParseError.pattern = 0;
        mParseError.character = 0;

        mEncoding = std::make_unique<Encoding>(nodes);
        NFA nfa(nodes, *mEncoding);
        mDFA = std::make_unique<DFA>(nfa, *mEncoding);
    }

    bool Matcher::valid() const
    {
        return (bool)mDFA;
    }

    const Matcher::ParseError &Matcher::parseError() const
    {
        return mParseError;
    }

    unsigned int Matcher::match(const std::string &string, unsigned int start, unsigned int &pattern) const
    {
        unsigned int state = mDFA->startState();
        unsigned int matched = 0;
        
        for(unsigned int i=start; i<string.size(); i++) {    
            Encoding::CodePoint codePoint = mEncoding->codePoint(string[i]);
            if(codePoint == Encoding::kInvalidCodePoint) {
                break;
            }
            unsigned int nextState = mDFA->transition(state, codePoint);
            
            if(nextState == mDFA->rejectState()) {
                break;
            } else {
                if(mDFA->accept(nextState, pattern)) {
                    matched = (i - start) + 1;
                }
                state = nextState;
            }
        }

        return matched;
    }

    unsigned int Matcher::numPatterns() const
    {
        return mNumPatterns;
    }
}

File: src/regex/matcher.rs
use crate::regex;
use regex::DFA;
use regex::Encoding;
use regex::NFA;
use regex::Parser;

pub struct ParseError {
    pub pattern: usize,
    pub pos: usize,
    pub message: String
}

impl ParseError {
    pub fn new(pattern: usize, pos: usize, message: String) -> ParseError {
        ParseError { pattern, pos, message }
    }
}
pub struct Matcher {
    dfa: DFA,
    encoding: Encoding
}

impl Matcher {
    pub fn new(patterns: Vec<String>) -> Result<Matcher, ParseError> {
        let mut nodes = Vec::new();
        for (i, pattern) in patterns.iter().enumerate() {
            match Parser::parse(pattern.as_str()) {
                Ok(node) => nodes.push(node),
                Err(error) => return Err(ParseError::new(i, error.pos, error.message))
            } 
        }

        println!("*** Parse nodes ***");
        for node in &nodes {
            node.print(0);
        }
        println!();

        let encoding = Encoding::new(&nodes);
        println!("*** Encoding ***");
        encoding.print();
        println!();

        let nfa = NFA::new(&nodes, &encoding);
        println!("*** NFA ***");
        nfa.print();
        println!();

        let dfa = DFA::new(&nfa, &encoding);
        println!("*** DFA ***");
        dfa.print();
        println!();

        return Ok(Matcher { dfa, encoding });
    }

    pub fn match_string(&self, string: &str, start: usize) -> (usize, usize) {
        let mut state = self.dfa.start_state;
        let mut num_matched = 0;
        let mut matched_pattern = usize::MAX;

        for i in start..string.len() {
            let code_point = self.encoding.code_point(string.chars().nth(i).unwrap());
            if code_point == regex::encoding::INVALID_CODE_POINT {
                break;
            }
            let next_state = self.dfa.transition(state, code_point);
            if next_state == self.dfa.reject_state {
                break;
            } else {
                if let Some(pattern) = self.dfa.accept(next_state) {
                    num_matched = (i - start) + 1;
                    matched_pattern = pattern;
                }
                state = next_state;
            }
        }

        return (num_matched, matched_pattern);
    }
}